lib/livebook_web/utils.ex

```diff
@@ -0,0 +1,34 @@
+defmodule LivebookWeb.Utils do
+  @moduledoc "Utility functions for web-focused tasks"
+
+  # Converts human-readable strings to strings which can be used
+  # as HTML element IDs (compatible with HTML4+)
+  #
+  # At the same time duplicate IDs are enumerated to avoid duplicates
+  @spec names_to_ids([binary], MapSet.t()) :: list
+  def names_to_ids(names, used_ids \\ MapSet.new())
+
+  def names_to_ids([], _), do: []
+
+  def names_to_ids([name | rest], used_ids) do
+    id = name_to_id(name)
+
+    id =
+      Stream.iterate(1, &(&1 + 1))
+      |> Stream.map(fn
+        1 -> id
+        index -> "#{id}-#{index}"
+      end)
+      |> Enum.find(&(!MapSet.member?(used_ids, &1)))
+
+    [id | names_to_ids(rest, MapSet.put(used_ids, id))]
```

Comment:
Note: for such recursive calls it's best to use tail recursion, so instead of `[elem | recursive_call(...)]`, we can do `recursive_call([elem | acc], ...)`, so that the compiler can optimise the recursion into iteration and we don't run out of memory for long lists.

---

Generally I think we could write this without brute forcing the names (here it doesn't matter much as we don't expect much duplications, but still). One idea I can think of:

```elixir
def names_to_ids(names) do
  {slugs, _} =
    names
    |> Enum.map(&name_to_slug/1)
    |> Enum.map_reduce(%{}, fn slug, frequencies ->
      frequencies = Map.update(frequencies, slug, 1, &(&1 + 1))
      frequency = frequencies[slug]

      if frequency == 1 do
        {slug, frequencies}
      else
        {"#{slug}-#{frequency}", frequencies}
      end
    end)

  slugs
end
```

Reply:
I knew that tail-call optimization was optimized, but honestly I'm still getting my head around it.  But I do agree that it's nicer to just use a helper function inside of the function.  It avoids the whole "hidden" second argument thing and the tail recursion thing nicely.

I've taken your suggestion but used `Stream.map/2` and `Stream.transform/3` to (hopefully) reduce memory and because `Stream.transform/3` has the nice property that it returns the new stream, so you don't need to do the `{slugs, _} =` and return `slug`.

Reply:
If you don't like matching on the tuple you can alternatively pipe into `|> elem(0)`.

Reply:
👍  I found that before reading your comment.  Thanks 😄

Reply:
> I knew that tail-call optimization was optimized, but honestly I'm still getting my head around it.

Just a piece of advice, I recommend reading SICP (Structure and implementation of computer programmes) it really helped me with recursion and functional programming in general.